File: 24_高级数据结构/good java/UndoDSU/UndoDSU.go
package main

import "fmt"

func main() {
	n := int32(10)
	dsu := NewUndoDSU(n)
	operation := dsu.Union(1, 2)
	operation.Apply()
	fmt.Println(dsu.Size(1) == 2)
	operation.Undo()
	fmt.Println(dsu.Size(1) == 1)
	fmt.Println(dsu.Size(2) == 1)
}

type IOperation interface {
	Apply()
	Undo()
}

type Operation struct{ apply, undo func() }

func NewOperation(apply, undo func()) *Operation { return &Operation{apply: apply, undo: undo} }
func (op *Operation) Apply()                     { op.apply() }
func (op *Operation) Undo()                      { op.undo() }

type UndoDSU struct {
	rank, parent []int32
}

func NewUndoDSU(n int32) *UndoDSU {
	rank := make([]int32, n)
	parent := make([]int32, n)
	for i := int32(0); i < n; i++ {
		rank[i] = 1
		parent[i] = -1
	}
	return &UndoDSU{rank: rank, parent: parent}
}

func (dsu *UndoDSU) Find(x int32) int32 {
	for dsu.parent[x] != -1 {
		x = dsu.parent[x]
	}
	return x
}

func (dsu *UndoDSU) Size(x int32) int32 {
	return dsu.rank[dsu.Find(x)]
}

func (dsu *UndoDSU) Union(a, b int32) IOperation {
	var x, y int32
	// to NamedTuple
	return NewOperation(
		func() {
			x = dsu.Find(a)
			y = dsu.Find(b)
			if x == y {
				return
			}
			if dsu.rank[x] < dsu.rank[y] {
				x, y = y, x
			}
			dsu.parent[y] = x
			dsu.rank[x] += dsu.rank[y]
		},
		func() {
			cur := y
			for dsu.parent[cur] != -1 {
				cur = dsu.parent[cur]
				dsu.rank[cur] -= dsu.rank[y]
			}
			dsu.parent[y] = -1
		},
	)
}

File: 24_高级数据结构/good java/UndoDSU/UndoDSU.java
package template.algo;

import java.util.Arrays;

public class UndoDSU {
    int[] rank;
    int[] p;

    public UndoDSU(int n) {
        rank = new int[n];
        p = new int[n];
    }

    public void init() {
        Arrays.fill(rank, 1);
        Arrays.fill(p, -1);
    }

    public int find(int x) {
        while (p[x] != -1) {
            x = p[x];
        }
        return x;
    }

    public int size(int x) {
        return rank[find(x)];
    }

    public UndoOperation merge(int a, int b) {
        return new UndoOperation() {
            int x, y;


            public void apply() {
                x = find(a);
                y = find(b);
                if (x == y) {
                    return;
                }
                if (rank[x] < rank[y]) {
                    int tmp = x;
                    x = y;
                    y = tmp;
                }
                p[y] = x;
                rank[x] += rank[y];
            }


            public void undo() {
                int cur = y;
                while (p[cur] != -1) {
                    cur = p[cur];
                    rank[cur] -= rank[y];
                }
                p[y] = -1;
            }
        };
    }
}
